source/extensions/filters/http/cache/cache_filter.h

```diff
@@ -80,21 +82,97 @@ class CacheFilter : public Http::PassThroughFilter,
 
   static LookupStatus resolveLookupStatus(absl::optional<CacheEntryStatus> cache_entry_status,
                                           FilterState filter_state);
 
 private:
+  class UpstreamRequest : public Http::AsyncClient::StreamCallbacks, public InsertQueueCallbacks {
+  public:
+    void sendHeaders(Http::RequestHeaderMap& request_headers);
+    void disconnectFilter();
+
+    // StreamCallbacks
+    void onHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override;
+    void onData(Buffer::Instance& data, bool end_stream) override;
+    void onTrailers(Http::ResponseTrailerMapPtr&& trailers) override;
+    void onComplete() override;
+    void onReset() override;
+
+    // InsertQueueCallbacks
+    void insertQueueOverHighWatermark() override;
+    void insertQueueUnderLowWatermark() override;
+    void insertQueueAborted() override;
+
+    static UpstreamRequest* create(CacheFilter* filter, std::shared_ptr<HttpCache> cache,
+                                   Http::AsyncClient& async_client,
+                                   const Http::AsyncClient::StreamOptions& options);
+    UpstreamRequest(CacheFilter* filter, std::shared_ptr<HttpCache> cache,
+                    Http::AsyncClient& async_client,
+                    const Http::AsyncClient::StreamOptions& options);
+    ~UpstreamRequest() override;
+
+  private:
+    // Precondition: lookup_result_ points to a cache lookup result that requires validation.
+    //               filter_state_ is ValidatingCachedResponse.
+    // Serves a validated cached response after updating it with a 304 response.
+    void processSuccessfulValidation(Http::ResponseHeaderMapPtr response_headers);
+
+    // Updates the filter state belonging to the UpstreamRequest, and the one belonging to
+    // the filter if it has not been destroyed.
+    void setFilterState(FilterState fs);
+
+    // Updates the insert status belonging to the filter, if it has not been destroyed.
+    void setInsertStatus(InsertStatus is);
+
+    // If an error occurs while the stream is active, abort will reset the stream, which
+    // in turn provokes the rest of the destruction process.
+    void abort();
+
+    // Precondition: lookup_result_ points to a cache lookup result that requires validation.
+    //               filter_state_ is ValidatingCachedResponse.
+    // Checks if a cached entry should be updated with a 304 response.
+    bool shouldUpdateCachedEntry(const Http::ResponseHeaderMap& response_headers) const;
+
+    CacheFilter* filter_ = nullptr;
+    LookupResultPtr lookup_result_;
+    LookupContextPtr lookup_;
+    bool is_head_request_;
+    bool request_allows_inserts_;
+    std::shared_ptr<const CacheFilterConfig> config_;
+    FilterState filter_state_;
+    std::shared_ptr<HttpCache> cache_;
+    Http::AsyncClient::Stream* stream_ = nullptr;
+    std::unique_ptr<UpstreamRequest> self_ownership_;
```

Comment:
I think we have linters that aren't big fans of bare pointers, plus *I* am not a fan of bare pointers (I think the ownership semantics are much clearer with explicitly `self_ownership`), plus in one of the later iterations it's going to become a smart pointer that *isn't* owned by itself so it's more convenient to not have it be a bare pointer temporarily and then have to rejigger it. :)

Reply:
I'm very sorry for the painfully slow replies--I should be done digging out from my problems elsewhere, and should be able to reply in a timely manner for the foreseeable future.

I didn't intend to suggest replacing anything with a bare pointer (though bare pointers are the right tool for many situations). My intent was to suggest
1. remove `std::unique_ptr<UpstreamRequest> self_ownership_;`
2. Replace the body of UpstreamRequest::create with `return new UpstreamRequest(...);`
3. Replace the body of UpstreamRequest::onReset, with `delete this`
4. In UpstreamRequest::onComplete, replace `self_ownership_ = nullptr` with `delete this`

I have several reasons for suggesting this:
1. The standard meaning of a `unique_ptr<Foo> foo_` member is "when this class is destroyed, that will in turn delete the pointer held by foo_." self_ownership_ inverts that meaning, which is counter to normal expectations.
2. `delete this` is a well-known pattern. In particular, it's reasonable to expect C++ programmers to know that it would be bad to touch any member variables after `delete this`. Even new C++ programmers should at least see `delete this` and think "something weird is happening." In comparison, `self_ownership_ = nullptr` just looks like a manipulation of a member, and doesn't raise any obvious danger flags. 
3. Getting rid of `self_ownership_` means one fewer member variable to understand, and simplifies the ownership structure.

If a future PR will introduce a real need for a smart pointer, I suspect it will be as convenient to add it then as to replace self_ownership_ with whatever it's to become.

Reply:
Done. Disagree but don't care that much.